// api-count/esm2015/lib/api-count.interceptor.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/api-count.interceptor.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { HttpResponse, HTTP_INTERCEPTORS } from '@angular/common/http';
import { throwError } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { LoaderService } from './loader.service';
import * as i0 from "@angular/core";
export class ApiCountInterceptor {
    /**
     * @param {?} injector
     */
    constructor(injector) {
        this.injector = injector;
    }
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    intercept(req, next) {
        /** @type {?} */
        const loaderService = this.injector.get(LoaderService);
        /** @type {?} */
        let cloneChanges = {};
        //Modify Request
        /** @type {?} */
        const authReq = req.clone(cloneChanges);
        loaderService.addAPI();
        return next.handle(authReq).pipe(map((/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            if (event instanceof HttpResponse) {
                loaderService.removeAPI();
            }
            return event;
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            /** @type {?} */
            let data = {};
            data = {
                reason: error && error.error && error.error.reason ? error.error.reason : '',
                status: error.status
            };
            loaderService.removeAPI();
            return throwError(error);
        })));
    }
}
ApiCountInterceptor.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ApiCountInterceptor.ctorParameters = () => [
    { type: Injector }
];
/** @nocollapse */ ApiCountInterceptor.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function ApiCountInterceptor_Factory() { return new ApiCountInterceptor(i0.ɵɵinject(i0.INJECTOR)); }, token: ApiCountInterceptor, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    ApiCountInterceptor.prototype.injector;
}
/** @type {?} */
export const httpInterceptorProviders = [
    { provide: HTTP_INTERCEPTORS, useClass: ApiCountInterceptor, multi: true },
];
