// api-count/esm2015/lib/loader.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/loader.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
export class LoaderService {
    constructor() {
        this.apiCount = 0;
        this.api = new BehaviorSubject(this.apiCount);
        this.apiCountObservable = this.api.asObservable();
    }
    /**
     * @return {?}
     */
    addAPI() {
        this.apiCount += 1;
        this.api.next(this.apiCount);
        console.log(`Ongoing API count ${this.apiCount}`);
    }
    /**
     * @return {?}
     */
    removeAPI() {
        this.apiCount -= 1;
        this.api.next(this.apiCount);
        console.log(`Ongoing API count ${this.apiCount}`);
    }
}
LoaderService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
LoaderService.ctorParameters = () => [];
/** @nocollapse */ LoaderService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function LoaderService_Factory() { return new LoaderService(); }, token: LoaderService, providedIn: "root" });
if (false) {
    /** @type {?} */
    LoaderService.prototype.apiCount;
    /**
     * @type {?}
     * @private
     */
    LoaderService.prototype.api;
    /** @type {?} */
    LoaderService.prototype.apiCountObservable;
}
